chainntnfs: handle spend notifications within TxConfNotifier

wpaulino opened the pull request:
In this PR, we overhaul the `TxConfNotifier` struct to also track spend notifications. This introduces a number of benefits, such as: removing duplicate spend notification logic within the different `ChainNotifier` implementations, spend reorg detection, and many other minor optimizations. This will serve as a stepping stone for re-enabling the height hint cache, as it previously had issues under reorgs. Details for the most important commits can be found inline, explaining what each one does and why.

Depends on #1787.

halseth commented on chainntnfs/tx_notifier.go:
```diff
@@ -1115,9 +1117,41 @@ func (n *TxNotifier) DisconnectTip(blockHeight uint32) error {
+	for spendHeight, ops := range n.opsBySpendHeight {
+		// We're only interesting in outpoints whose spending
+		// transaction was included at the height being disconnected.
+		if spendHeight != blockHeight {
```
just index `n.opsBySpendHeight[blockHeight]` instead of looping?

wpaulino replied:
Indeed - don't know what I was thinking. Fixed!

halseth commented on chainntnfs/tx_notifier.go:
```diff
@@ -1164,6 +1198,38 @@ func (n *TxNotifier) dispatchConfReorg(ntfn *ConfNtfn, heightDisconnected uint32
+	// cahnnel are always non-blocking.
+	select {
+	case <-ntfn.Event.Spend:
+	case <-n.quit:
```
nit: selecting on quit not needed here when we have default clause.

wpaulino replied:
Fixed.

halseth commented on chainntnfs/tx_notifier.go:
```diff
@@ -1164,6 +1198,38 @@ func (n *TxNotifier) dispatchConfReorg(ntfn *ConfNtfn, heightDisconnected uint32
+	default:
+	}
+
+	// Send a reorg notifiaction to the client in order for them to
```
nit: typo

wpaulino replied:
Fixed.

halseth commented on chainntnfs/tx_notifier.go:
```diff
@@ -1156,6 +1105,105 @@ func (n *TxNotifier) DisconnectTip(blockHeight uint32) error {
+// NOTE: This must be called with the TxNotifier's lock held and after it's
+// height has already been reflected by a block being connected/disconnected.
+func (n *TxNotifier) updateHints(disconnect bool) {
+	// In order to retrieve the corect transactions and outpoints for which
```
nit: typo

wpaulino replied:
Fixed.

halseth commented on chainntnfs/tx_notifier.go:
```diff
@@ -1156,6 +1105,105 @@ func (n *TxNotifier) DisconnectTip(blockHeight uint32) error {
+	// was disconnected or not. If it was, then we'll need to look at the
+	// next height as the TxNotifier's height has already reflected the
+	// block being disconnected.
+	confirmSpendHeight := n.currentHeight
```
can the height be passed into the method, instead of doing this?

wpaulino replied:
Yeah, definitely much cleaner solution. Fixed!

halseth commented on lnd_test.go:
```diff
@@ -6260,14 +6282,9 @@ func testRevokedCloseRetributionZeroValueRemoteOutput(net *lntest.NetworkHarness
-	if *txid != *closeTxId {
-		t.Fatalf("expected closeTx(%v) in mempool, instead found %v",
-			closeTxId, txid)
+	if !waitForTxidInMempool(net.Miner.Node, closeTxId, 20*time.Second) {
```
Will the sweep broadcast be deterministic? If that's the case I feel we should wait for two txes instead.

wpaulino replied:
It will. The flake has only seem to come up with these changes introduced. Should just be a timing issue - but I decided to go with this approach based on the comment above wanting to ensure that the transaction was broadcast.

halseth replied:
Why not wait for both txes then?

wpaulino replied:
This wasn't needed after all. The counterparty shouldn't act upon mempool spends. Not sure how this happened, but haven't been able to reproduce it after many runs locally and on Travis.

wpaulino replied:
Turns out the flake I was seeing has already been addressed in https://github.com/lightningnetwork/lnd/pull/1621/commits/90f9fa2c14105cf06eae326a09d845098b178e4e#diff-61b0847f8902e7a5d746ba18a02649d5R12292.

wpaulino commented on chainntnfs/tx_notifier.go:
```diff
@@ -784,13 +793,57 @@ func (n *TxNotifier) ConnectTip(blockHash *chainhash.Hash,
 	for _, tx := range txns {
+		// In order to determine if this transaction is relevant to the
+		// notifier, we'll check its inputs for any outstanding spend
+		// notifications.
```
Fixed.

wpaulino commented on chainntnfs/tx_notifier.go:
```diff
@@ -555,6 +555,210 @@ func (n *TxNotifier) dispatchConfDetails(
+//
+// The registration succeeds if no errors is returned. If the returned
+// HistoricalSpendDisaptch is non-nil, the caller is responsible for attempting
+// determine whether the outpoint has been spent between the start and end
```
Fixed.

wpaulino commented on chainntnfs/tx_notifier.go:
```diff
@@ -555,6 +555,210 @@ func (n *TxNotifier) dispatchConfDetails(
+// RegisterSpend handles a new spend notification request. The client will be
+// notified once the outpoint is detected as spent within the chain.
+//
+// The registration succeeds if no errors is returned. If the returned
```
Fixed.

wpaulino commented on chainntnfs/tx_notifier.go:
```diff
@@ -77,28 +152,82 @@ type HistoricalConfDispatch struct {
+	// notification for.
+	OutPoint wire.OutPoint
+
+	// PkScript ...
```
Fixed.

wpaulino commented on chainntnfs/tx_notifier.go:
```diff
@@ -15,6 +16,80 @@ var (
 
+// rescanState indicates the progression of a registration before the notifier
+// can begin dispatching confirmations at tip.
+type rescanState byte
```
Fixed.

wpaulino commented on chainntnfs/interface.go:
```diff
@@ -196,14 +206,28 @@ type SpendDetail struct {
+	// Reorg is a channel that will be sent upon once we detect the spending
+	// transaction of the outpoint in question has been reorged out of the
+	// chain.
+	Reorg chan struct{}
```
Which part?

halseth replied:
If a spend is registered, and the spend notification is sent, who will read the `Reorg` channel if it gets reorged out? Feels like something that should be left to the conf notifications.

wpaulino replied:
Just like with confirmation notifications, it's left to the caller. The same can be said for those as we don't read the `NegativeConf` channel anywhere throughout the codebase.

wpaulino replied:
Fixed.

wpaulino commented on chainntnfs/interface.go:
```diff
@@ -196,14 +206,28 @@ type SpendDetail struct {
 	// Spend is a receive only channel which will be sent upon once the
 	// target outpoint has been spent.
-	Spend <-chan *SpendDetail // MUST be buffered.
+	Spend chan *SpendDetail // MUST be buffered.
```
This had to be modified when moving the logic to the `TxConfNotifier` in order to dispatch the spending details. Previously, we were declaring the spend channel outside of the `SpendEvent` creation and passing it in. We could do that here, but then we'd need to track the channels, rather than having them be a part of the notification struct. The extra code to accomplish doesn't seem worth it to me.

wpaulino commented on chainntnfs/height_hint_cache.go:
```diff
@@ -127,6 +127,10 @@ func (c *HeightHintCache) CommitSpendHint(height uint32, ops ...wire.OutPoint) e
 		return nil
 	}
 
+	if len(ops) == 0 {
```
It's possible in the event that we don't need to update the hints for any transactions/outpoints. In that case, rather than handling the check there, I figured it'd be better to have it within the cache itself to avoid the extra database hit.

wpaulino commented:
Addressed remaining comments and extended the tests a bit more to correctly test the behavior of the `SpendHintCache` like we do with the `ConfirmHintCache`.

halseth commented on chainntnfs/btcdnotify/btcd.go:
```diff
@@ -464,6 +430,8 @@ out:
 
 			spendingTx := newSpend.tx
+			spendingTxHash := newSpend.tx.Hash()
+			activeSpendNtfns := b.txNotifier.ActiveSpendNotifications()
```
same as for bitcoind

halseth commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -402,40 +424,34 @@ func (b *BitcoindNotifier) handleRelevantTx(tx chain.RelevantTx, bestHeight int3
-	for i, txIn := range msgTx.TxIn {
+	spendingTx := tx.TxRecord.MsgTx
+	spendingTxHash := spendingTx.TxHash()
+	activeSpendNtfns := b.txNotifier.ActiveSpendNotifications()
```
The diff is gone now, but I think it could be racy since after you get the active spend notifications, you have no guarantee no more are added to the active set while using those you fetched.

halseth commented on chainntnfs/txnotifier.go:
```diff
@@ -872,6 +925,7 @@ func (n *TxNotifier) ConnectTip(blockHash *chainhash.Hash,
 			TxIndex:     uint32(tx.Index()),
 		}
 
+		// TODO(wilmer): cancel pending historical rescans if any?
```
👍

halseth commented on chainntnfs/txnotifier.go:
```diff
@@ -1162,9 +1164,34 @@ func (n *TxNotifier) DisconnectTip(blockHeight uint32) error {
+
+		// For all requests which have had a spend notification
+		// dispatched, we'll attempt to drain it and send a reorg
+		// notifiaction instead.
```
nit: typo

halseth commented on chainntnfs/txnotifier.go:
```diff
@@ -1196,6 +1145,95 @@ func (n *TxNotifier) DisconnectTip(blockHeight uint32) error {
+// determine which transactions and outpoints we should update based on whether
+// a new block is being connected/disconnected.
+//
+// NOTE: This must be called with the TxNotifier's lock held and after it's
```
nit: s/it's/its

halseth commented on chainntnfs/txnotifier.go:
```diff
@@ -1196,6 +1145,95 @@ func (n *TxNotifier) DisconnectTip(blockHeight uint32) error {
+		// it as we can't yet determine if it's confirmed or not.
+		if confNtfnSet.rescanStatus != rescanComplete ||
+			confNtfnSet.details != nil {
+
```
nit: remove newline

halseth commented on chainntnfs/txnotifier.go:
```diff
@@ -1196,6 +1145,95 @@ func (n *TxNotifier) DisconnectTip(blockHeight uint32) error {
+		// can't yet determine if it's unspent or not.
+		if spendNtfnSet.rescanStatus != rescanComplete ||
+			spendNtfnSet.details != nil {
+
```
same

halseth commented on chainntnfs/txnotifier_test.go:
```diff
@@ -413,16 +417,498 @@ func TestTxConfHistoricalDispatch(t *testing.T) {
+			t.Fatalf("request #%d expected to received spend "+
+				"notification", i)
+		}
+	}
```
Good test! 💯 

Attempt to register a new client after update spend details was called, that we expect to get notified?

halseth commented on chainntnfs/txnotifier_test.go:
```diff
@@ -413,16 +417,498 @@ func TestTxConfHistoricalDispatch(t *testing.T) {
+	// Before extending the notifier's tip with the dummy block above, we'll
+	// cancel the second request.
+	n.CancelSpend(ntfn2.OutPoint, ntfn2.SpendID)
+	err := n.ConnectTip(block.Hash(), startingHeight+1, block.Transactions())
```
nit: add newline above

halseth commented on chainntnfs/txnotifier_test.go:
```diff
@@ -679,18 +1165,221 @@ func TestTxConfChainReorg(t *testing.T) {
+	}
+
+	ntfn2 := &chainntnfs.SpendNtfn{
+		SpendID:  21,
```
good choice of spend ID 👍

halseth commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -717,167 +705,118 @@ type spendCancel struct {
-	// chain extends and the output remains unspent.
 	if txOut != nil {
-		err := b.spendHintCache.CommitSpendHint(heightHint, *outpoint)
+		err := b.txNotifier.UpdateSpendDetails(*outpoint, nil)
```
Should have a comment.

halseth commented on chainntnfs/interface_test.go:
```diff
@@ -16,25 +16,17 @@ import (
-	"github.com/lightningnetwork/lnd/channeldb"
-
-	// Required to auto-register the bitcoind backed ChainNotifier
+	"github.com/lightningnetwork/lnd/chainntnfs/bitcoindnotify" // Required to auto-register the btcd backed ChainNotifier
```
messed up import comments?

halseth commented on chainntnfs/interface_test.go:
```diff
@@ -303,11 +295,7 @@ func testSpendNotification(miner *rpctest.Harness,
-	// mempool, we use a quite large timeout of 10 seconds.
-	// TODO(halseth): change this when mempool spends are removed.
-	mempoolSpendTimeout := 10 * time.Second
+	mempoolSpendTimeout := 2 * chainntnfs.TrickleInterval
```
can potentially be made a separate PR.

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -557,6 +557,272 @@ func (n *TxNotifier) dispatchConfDetails(
+	spendSet, ok := n.spendNotifications[op]
+	if !ok {
+		return fmt.Errorf("no notifications found for outpoint %v", op)
+	}
```
Fixed.

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -557,6 +557,272 @@ func (n *TxNotifier) dispatchConfDetails(
+	// been spent.
+	case rescanNotStarted:
+	}
+
```
Fixed.

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -557,6 +557,272 @@ func (n *TxNotifier) dispatchConfDetails(
+	// We'll now let the caller know whether a historical rescan is needed
+	// depending on the current rescan status.
+	switch spendSet.rescanStatus {
+	// If the spending details for this outpoint have already been
```
Fixed.

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -177,17 +268,23 @@ type TxNotifier struct {
+// dispatching a recan for a historical event in the chain.
 func NewTxNotifier(startHeight uint32, reorgSafetyLimit uint32,
-	confirmHintCache ConfirmHintCache) *TxNotifier {
+	confirmHintCache ConfirmHintCache,
```
Agreed - we already have the unified `HeightHintCache` struct that handles both of these cases. Perhaps this should be done within #2044 or even another PR?

cfromknecht replied:
just remembering that it is already done in, though it was never merged https://github.com/lightningnetwork/lnd/pull/1767

wpaulino commented on chainntnfs/neutrinonotify/neutrino.go:
```diff
@@ -941,8 +922,49 @@ func (n *NeutrinoNotifier) RegisterConfirmationsNtfn(txid *chainhash.Hash,
+	// event handler and wait for its response.
+	errChan := make(chan error, 1)
+	select {
+	case n.notificationRegistry <- &rescanFilterUpdate{
```
Fixed.

cfromknecht commented on chainntnfs/interface.go:
```diff
@@ -196,14 +206,30 @@ type SpendDetail struct {
 	// Spend is a receive only channel which will be sent upon once the
 	// target outpoint has been spent.
-	Spend <-chan *SpendDetail // MUST be buffered.
+	Spend chan *SpendDetail // MUST be buffered.
```
would prefer the `MUST be buffered` comment to be above so it is properly captured in godocs

wpaulino replied:
Fixed.

cfromknecht commented on chainntnfs/txnotifier.go:
```diff
@@ -557,6 +557,300 @@ func (n *TxNotifier) dispatchConfDetails(
+	startHeight := ntfn.HeightHint
+	hint, err := n.spendHintCache.QuerySpendHint(ntfn.OutPoint)
+	if err == nil {
+		if hint > startHeight {
```
those execution paths wouldn't be equivalent. i think as is it's correct

wpaulino replied:
Agree with @cfromknecht.

halseth replied:
In this case it would be equivalent since the `else if` would only be hit in the case `err != nil`, but this is fine.

cfromknecht replied:
if the first check is `err == nil && hint > startHeight`, then `else if` can be hit when `err == nil`, which would trigger the log message

halseth replied:
Oh yeah, you're right 🤦‍♂️

cfromknecht commented on chainntnfs/txnotifier.go:
```diff
@@ -1243,6 +1270,36 @@ func (n *TxNotifier) dispatchConfReorg(ntfn *ConfNtfn,
+	}
+
+	ntfn.dispatched = false
+	n.spendNotifications[ntfn.OutPoint].details = nil
```
this is already done before calling `dispatchSpendReorg`, no?

wpaulino replied:
Yep, removed from here.

cfromknecht commented on chainntnfs/txnotifier.go:
```diff
@@ -1227,6 +1176,93 @@ func (n *TxNotifier) DisconnectTip(blockHeight uint32) error {
+//
+// NOTE: This must be called with the TxNotifier's lock held and after its
+// height has already been reflected by a block being connected/disconnected.
+func (n *TxNotifier) updateHints(height uint32) {
```
nice refactor, really nice that both of these are being handled in one place. down the road i think it'd be worth considering updating both types of hints under a single db transaction, but we can cross that bridge when we get there

wpaulino replied:
Indeed - added a `TODO`.

cfromknecht commented on chainntnfs/bitcoindnotify/bitcoind.go:
```diff
@@ -901,38 +843,32 @@ func (b *BitcoindNotifier) dispatchSpendDetailsManually(op wire.OutPoint,
+				err := b.txNotifier.UpdateSpendDetails(
+					op, details,
+				)
+				if err != nil {
```
won't this continue to scan blocks after the spend was found? seems like we should return the result of `UpdateSpendDetails` directly

wpaulino replied:
A1 catch! Perhaps this was intentional 😉

wpaulino commented on chainntnfs/interface_test.go:
```diff
@@ -1108,6 +1112,156 @@ func testReorgConf(miner *rpctest.Harness, notifier chainntnfs.TestChainNotifier
+
+	// We disconnect the two nodes, such that we can start mining on them
+	// individually without the other one learning about the new blocks.
+	err = miner.Node.AddNode(miner2.P2PAddress(), rpcclient.ANRemove)
```
> I think we can actually just use the Node line of RPC's?

Isn't that what it's already using? Or are you referring to something else?

halseth replied:
You can see examples of using the new Node API here: https://github.com/lightningnetwork/lnd/pull/1621/commits/7e75586ed0b2577a2be0528d64c5703d4a867821

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -792,6 +806,17 @@ func (n *TxNotifier) updateSpendDetails(op wire.OutPoint,
 	// If the historical rescan was not able to find a spending transaction
 	// for this outpoint, then we can track the spend at tip.
 	if details == nil {
+		// We'll commit the current height as the spend hint to prevent
```
It's now been moved here to aggregate them all in one place.

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -1031,17 +1085,33 @@ out:
+		for op := range n.opsBySpendHeight[matureBlockHeight] {
+			delete(n.spendNotifications, op)
+		}
+		delete(n.opsBySpendHeight, matureBlockHeight)
```
I don't think we want to remove the height hint here actually as it _could_ lead to prematurely removing it.

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -876,14 +885,58 @@ func (n *TxNotifier) ConnectTip(blockHash *chainhash.Hash,
+				SpendingHeight:    int32(blockHeight),
+			}
+
+			// TODO(wilmer): cancel pending historical rescans if any?
```
> If early cancellation then, should be straight forward to thread thru a cancel channel within the spend set itself. Depends on how the backend is handling the rescan though.

Yep this `TODO` is referring to cancelling the pending rescan in the event that we see it at tip. Can be left as a follow-up IMO.

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -851,13 +851,22 @@ func (n *TxNotifier) dispatchSpendDetails(ntfn *SpendNtfn, details *SpendDetail)
+// notification will be dispatched to the relevant clients. Confirmation
+// notifications will only be dispatched for transactions that have met the
+// required number of confirmations required by the client.
+func (n *TxNotifier) ConnectTip(blockHash *chainhash.Hash, blockHeight uint32,
```
Yup!

wpaulino commented on chainntnfs/txnotifier.go:
```diff
@@ -557,6 +557,300 @@ func (n *TxNotifier) dispatchConfDetails(
+// check whether the transaction is relevant to the notifier if it spends any
+// outputs for which we currently have registered notifications for. If it is
+// relevant, spend notifications will be dispatched to the caller.
+func (n *TxNotifier) ProcessRelevantTx(tx *wire.MsgTx, txHeight int32) error {
```
Should be spends only - renamed to `ProcessRelevantSpendTx`.

wpaulino commented on chainntnfs/interface.go:
```diff
@@ -174,6 +174,16 @@ type ConfirmationEvent struct {
+// NewConfirmationEvent constructs a new ConfirmationEvent with newly opened
+// channels.
+func NewConfirmationEvent(numConfs uint32) *ConfirmationEvent {
+	return &ConfirmationEvent{
```
Yep this can be done as well. Can be done as a follow-up IMO.
